netatop-0.6/man/netatopd.8
.TH NETATOPD 8 "October 2012" "Linux"
.SH NAME
.B netatopd
- log network statistics of finished processes
.SH DESCRIPTION
This manual page documents the 
.I netatopd
daemon. This daemon reads the counters of exited processes
delivered by the
.I netatop
module and writes the counters to a logfile.
.PP
The 'usual' state of this daemon is the getsockopt()
command NETATOP_GETCNT_EXIT. With this command,
.I netatopd
blocks until an exited task is available on the exitlist of the 
.I netatop
module. The obtained netpertask struct that countains the
counters of an exited process is written to a logfile
(similar to the process accounting info provided by the base kernel
itself).
The daeon stops logging when there is only 5% of free space left
in the filesystem of the logfile.
.PP
The logfile starts with a small header struct that contains a.o. a sequence
number indicating how many netpertask structs are currently logged.
This header is mmapped and can be consulted by analysis processes like
.I atop
that consult the logfile.
Behind the header, the netpertask structs can be found in compressed
form. Every compressed netpertask struct is preceeded by one byte 
specifying the size of the compressed netpertask struct.
.PP
Before an analysis process starts using the logfile, it has to subscribe
itself by decrementing a semaphore owned by the
.I netatopd
daemon. In this way,
.I netatopd
knows how many analysis processes are currently using the logfile.
When no processes use the logfile any more,
.I netatopd
truncates the logfile and start building a new one as soon as
a new subscription of an analysis process is noticed.
.PP
.SH FILES
.PP
.TP 5
.B /var/run/netatop.log
File in which 
.I netatopd
writes the compressed accounting records about 
processes that have finished.
.SH SEE ALSO
.B netatop(4),
.B atop(1),
.B atopsar(1),
.B atoprc(5)
.br
.B http://www.atoptool.nl
.SH AUTHOR
Gerlof Langeveld (gerlof.langeveld@atoptool.nl)


netatop-0.6/ChangeLog
commit eba568b369605d4d525b05e0ae8f4faaa7a6466f
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Tue Aug 4 12:06:31 2015 +0200

    Bug solution:
    Arguments of the hook function have been modified in kernel version 4.1
    resulting in a kernel panic.
    Credits: David B. Lamkins

M	module/netatop.c

commit 01e87c5384c60188a419d3f37bb44c9a64ed2812
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Sun Jan 18 14:15:21 2015 +0100

    For kernel version >= 3.17, a 64-bit division fails on 32-bit systems.
    The division is done now with div_64u().

M	module/netatop.c

commit cf0d3ea78fc87d938e6a15f3332912be9e43d252
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Sat Nov 22 16:27:22 2014 +0100

    The type of the first calling argument of the function
    typedef nf_hookfn has been changed from an 'unsigned int'
    to a 'const struct nf_hook_ops *' in kernel version 3.10
    (or 3.13?). Kernel-version dependent code has been added.

M	module/netatop.c

commit 5481f03746c3e7e615032dc656601a353c012306
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Sat Nov 22 15:57:35 2014 +0100

    The member 'real_start_time' in the task_struct has been modified
    from a 'struct timespec' to a 'u64' (in nanoseconds) in kernel 3.17.
    Kernel-version dependent code has been added to handle this.

M	module/netatop.c

commit a159daa73e53e0717887b687763d7cca5f7d23de
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Fri Feb 7 10:38:21 2014 +0100

    Added zlib1-dev for Debian derivatives.

M	README

commit f72c6ae6b7f3b2efbbfae97d25674e046f472d1f
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Fri Jul 19 08:43:12 2013 +0200

    Add dependency of zlib-devel package.

M	README

commit 8a0a7707c407e6e361b9f9475adc923f7d478c02
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Fri Jul 19 08:42:29 2013 +0200

    iUse install instead of cp command to avoid "text file busy".

M	Makefile

commit 6a544ec9c0fd5054489f9549b34de447deac5513
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Fri Jul 19 08:41:11 2013 +0200

    Convert /proc interface to proc_create/seq_file.
    Credits: Gabriel de Perthuis

M	module/netatop.c

commit a530702f4d6cf8607419aec671bc1cbf32e69dba
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Wed Dec 19 08:23:26 2012 +0100

    Show module version in /proc/netatop.

M	module/netatop.c

commit 1acce78c015e9825cb62761cd87f6ec5f4d6aba3
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Wed Dec 19 08:12:19 2012 +0100

    Use mutex instead of spinlock to protect the garbage collector
    from being called in parallel.

M	module/netatop.c

commit 0d8b252e5345a2d11aa934e2a4635e377423e006
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Mon Dec 17 15:32:49 2012 +0100

    Addition of the kernel thread 'knetatopgc' to issue the
    garbage collection periodically. This solves a panic
    situation (mainly on virtual servers) where function find_vpid
    is called in the context of a thread that is just busy to exit.

M	module/netatop.c

commit 3ae2b983eed4c3b2025593290382eb8bee2939e4
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Tue Dec 4 14:05:50 2012 +0100

    Optimize memory usage by creating slab caches for sockinfo and
    taskinfo structs, and by increasing the space allowed for
    sockinfo structs (appeared to be too tight for large servers).

M	module/netatop.c

commit 92512f25ff88db0b1cbc9fbd7109e8655628b6bb
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Mon Nov 19 10:50:41 2012 +0100

    Implemented separate functions for start and stop in init-script.

M	README
M	netatop.init

commit e51ab1c074f8a27cca8cb8e40014283ca27f10ba
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Mon Nov 19 10:49:07 2012 +0100

    Increased maximum space for taskinfo structs to 2MiB (was 1MiB) and
    increased counters in /proc/netatop from 9 till 12 positions.

M	module/netatop.c

commit 38698454b44f873f8895cad17306d5ad288fb2dc
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Mon Nov 12 13:24:53 2012 +0100

    Maintain an overflow counter for sockinfo and taskinfo structs that
    were omitted due to reaching their memory thresholds.

M	module/netatop.c
D	netatopversion.h

commit ff821a1124451d205ef84c02d17cebd425df6a77
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Sat Nov 10 16:29:17 2012 +0100

    Separate spin-locks have been added for the global varibales nrs and nrt
    due to failed modifications on these variables (specifically nrs).

M	module/netatop.c
M	netatopversion.h

commit a07903439c87c44d66aa3424610905c7e406c7e1
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Fri Oct 26 10:34:37 2012 +0200

    Added Changelog to archive.

M	mkdistr
M	netatopversion.h

commit 02e360e54426229afc0c30a2d0b9f751ff934b05
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Fri Oct 26 09:18:47 2012 +0200

    Adapt for proper versioning.

M	mkdistr
M	netatopversion.h

commit 5f849515d0e0fc66c9dfe2ab36abd482aa4b95f1
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Thu Oct 25 23:21:18 2012 +0200

    emoved automatic chkconfig.

M	Makefile

commit daa8edfd29db9f7c3ed07b24b4c66085c8fdb9e7
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Tue Oct 23 14:45:05 2012 +0200

    Make it possible for atop to dynamically detect netatop(d) per sample.

M	Makefile
M	README
M	daemon/netatopd.c
M	mkdistr
M	module/netatop.c
M	netatop.h
M	netatopd.h
M	netatopversion.h

commit a500af96306856e4d7dbb29e2ba387a3f592f7ba
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Mon Oct 22 20:52:13 2012 +0200

    Make distribution ready and add man-pages.

M	Makefile
A	man/netatop.4
A	man/netatopd.8
M	mkdistr
M	netatopversion.h

commit 3c6edf2395dd9b6386cb9f0002a79372a69903c4
Author: Gerlof Langeveld <gerlof@atcomputing.nl>
Date:   Sun Oct 14 15:25:25 2012 +0200

    Bug- and porting-solutions.

M	Makefile
A	README
M	daemon/Makefile
M	daemon/netatopd.c
A	mkdistr

netatop-0.6/daemon/netatopd.c
/*
** The netatopd daemon can be started optionally to read the counters
** of exited processes from the netatop module. The 'usual' state of this
** daemon is the getsockopt command NETATOP_GETCNT_EXIT. With this command
** netatopd blocks until an exited task is available on the exitlist of the 
** netatop module. The obtained netpertask struct is written to a logfile
** (similar to the process accounting info provided by the base kernel
** itself) when there is at least 5% of free space in the filesystem of
** the logfile.
**
** The logfile contains a small header struct that contains a.o. a sequence
** number indicating how many netpertask structs are currently logged.
** This header is mmapped and can be consulted by analysis processes that
** consult the logfile (e.g. atop).
** Behind the header, the netpertask structs can be found in compressed
** form. Every compressed netpertask struct is preceeded by one byte 
** specifying the size of the compressed netpertask struct.
**
** Before an analysis process starts using the logfile, it has to subscribe
** itself by decrementing the second semaphore of a semaphore group owned
** by netatopd. In this way netatopd knows how many analysis processes
** are using the logfile. When no processes use the logfile any more,
** netatopd removes the logfile and start building a new one as soon as
** a new subscription of an analysis process is noticed.
** ----------------------------------------------------------------------
** Copyright (C) 2012    Gerlof Langeveld (gerlof.langeveld@atoptool.nl)
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License version 2 as
** published by the Free Software Foundation.
*/
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <syslog.h>
#include <string.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ipc.h>
#include <sys/sem.h>
#include <errno.h>
#include <zlib.h>
#include <time.h>
#include <signal.h>
#include <sys/mman.h>
#include <sys/statvfs.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>

#include "../netatop.h"
#include "../netatopd.h"
#include "../netatopversion.h"

#define	EVER	;;

#include <stdio.h>
#include <string.h>

static int	histopen(struct naheader **);
static void	recstore(int, struct netpertask *, socklen_t);

/*
** Semaphore-handling
**
** A semaphore-group with two semaphores is created. The first semaphore
** specifies the number of netatopd processes running (to be sure that only
** one daemon is active at the time) ) and the second reflects the number
** of processes using the log-file (inverted).
** This second semaphore is initialized at some high value and is
** decremented by every analysis process (like atop) that uses the log-file
** and incremented as soon as such analysis process stops again.
*/
#define SEMTOTAL        100
#define	NUMCLIENTS	(SEMTOTAL - semctl(semid, 1, GETVAL, 0))

int
main(int argc, char *argv[])
{
	int 			i, netsock, histfd = -1, semid;
	struct netpertask	npt;
	socklen_t 		len = sizeof npt;
	struct naheader 	*nap;
	struct sigaction        sigact;
	void			gethup(int);
	struct sembuf		semincr = {0, +1, SEM_UNDO};

	/*
	** version number required?
	*/
	if (argc == 2 && *argv[1] == '-' && *(argv[1]+1) == 'v')
	{
		printf("%s - %s     <gerlof.langeveld@atoptool.nl>\n",
			NETATOPVERSION, NETATOPDATE);
		return 0;
	}

	/*
 	** verify if we are running with the right privileges
	*/
	if (geteuid() != 0)
	{
		fprintf(stderr, "Root privileges are needed!\n");
		exit(1);
	}

	/*
 	** open socket to IP layer 
	*/
	if ( (netsock = socket(PF_INET, SOCK_RAW, IPPROTO_RAW)) == -1)
	{
		perror("open raw socket");
		exit(2);
	}

	/*
	** create the semaphore group and initialize it;
	** if it already exists, verify if a netatopd daemon
	** is already running
	*/
	if ( (semid = semget(SEMAKEY, 0, 0)) >= 0)	// exists?
	{
		if ( semctl(semid, 0, GETVAL, 0) == 1)
		{
			fprintf(stderr, "Another netatopd is already running!");
			exit(3);
		}
	}
	else
	{
		if ( (semid = semget(SEMAKEY, 2, 0600|IPC_CREAT|IPC_EXCL)) >= 0)
		{
			(void) semctl(semid, 0, SETVAL, 0);
			(void) semctl(semid, 1, SETVAL, SEMTOTAL);
		}
		else
		{
			perror("cannot create semaphore");
			exit(3);
		}
	}

	/*
	** daemonize this process
	*/
	if ( fork() )
		exit(0);	// implicitly switch to background

	setsid();

	if ( fork() )
		exit(0);

	for (i=0; i < 1024; i++)
		if (i != netsock)
			close(i);

	umask(022);

	chdir("/tmp");

	/*
	** open syslog interface for failure messages
	*/
	openlog("netatopd", LOG_PID, LOG_DAEMON);
	syslog(LOG_INFO, "version %s actived\n", NETATOPVERSION);

	/*
	** raise semaphore to define a busy netatopd
	*/
	if ( semop(semid, &semincr, 1) == -1)
       	{
		syslog(LOG_ERR, "cannot increment semaphore\n");
		exit(3);
	}

	/*
	** the daemon can be woken up from getsockopt by receiving 
	** the sighup signal to verify if there are no clients any more
	** (truncate exitfile)
	*/
        memset(&sigact, 0, sizeof sigact);
        sigact.sa_handler = gethup;
        sigaction(SIGHUP, &sigact, (struct sigaction *)0);

	/*
	** raise priority
	*/
	(void) nice(-39);

	/*
	** open history file
	*/
	histfd = histopen(&nap);

	/*
	** continuously obtain the info about exited processes
	*/
	for (EVER)
	{
		/*
		** check if anybody is interested in the exitfile
		** if not, close and truncate it
		*/
		if (NUMCLIENTS == 0 && nap->curseq != 0)
		{
			/*
			** destroy and reopen history file
			*/
			munmap(nap, sizeof(struct naheader));
			close(histfd);
			syslog(LOG_INFO, "reopen history file\n");
			histfd = histopen(&nap);
		}

		/*
 		** get next exited process (call blocks till available)
		*/
		switch (getsockopt(netsock, SOL_IP, NETATOP_GETCNT_EXIT,
								&npt, &len))
		{
		   case 0:
			// skip if nobody is using it
			if (NUMCLIENTS == 0)
				continue;

			/*
			** at least one listener is active, so
			** store record for exited process
			*/
			recstore(histfd, &npt, len);

			/*
			** increment sequence number in file header
			*/
			nap->curseq++;
			break;

		   default:	// getsockopt failed
			switch (errno)
			{
				// no netatop module loaded?
			   case ENOPROTOOPT:
				sleep(10);
				continue;

				// signal received?
			   case EINTR:
				continue;

			   default:
				syslog(LOG_ERR, "getsockopt failed\n");
				return 0;
			}
		}
	}

	return 0;
}

/*
** open history file
*/
static int
histopen(struct naheader **nahp)
{
	int			fd;
	struct naheader 	nahdr = {MYMAGIC, 0,
					sizeof(struct naheader),
		                	sizeof(struct netpertask),
					getpid()};
	/*
 	** remove the old file; this way atop can detect that a
	** new file must be opened
	*/
	(void) unlink(NETEXITFILE);

	/*
 	** open new file
	*/
	if ( (fd = open(NETEXITFILE, O_RDWR|O_CREAT|O_TRUNC, 0644)) == -1)
	{
		syslog(LOG_ERR, "cannot open %s for write\n", NETEXITFILE);
		exit(3);
	}

	/*
 	** write new header and mmap
	*/
	if ( write(fd, &nahdr, sizeof nahdr) != sizeof nahdr)
	{
		syslog(LOG_ERR, "cannot write to %s\n", NETEXITFILE);
		exit(3);
	}

	*nahp = mmap((void *)0, sizeof *nahp, PROT_WRITE, MAP_SHARED, fd, 0);

	if (*nahp == (void *) -1)
	{
		syslog(LOG_ERR, "mmap of %s failed\n", NETEXITFILE);
		exit(3);
	}

	return fd;
}

static void
recstore(int fd, struct netpertask *np, socklen_t len)
{
	Byte		compbuf[sizeof *np + 128];
	unsigned long	complen = sizeof compbuf -1;
	struct statvfs	statvfs;
	int		rv;

	/*
 	** check if the filesystem is not filled for more than 95%
	*/
	if ( fstatvfs(fd, &statvfs) != -1)
	{
		if (statvfs.f_bfree * 100 / statvfs.f_blocks < 5)
		{
			syslog(LOG_ERR, "Filesystem > 95%% full; "
			                "write skipped\n");
			return;
		}
	}

	/*
 	** filesystem space sufficient
	** compress netpertask struct
	*/
	rv = compress(compbuf+1, &complen, (Byte *)np,
					(unsigned long)sizeof *np);
	switch (rv)
	{
           case Z_OK:
           case Z_STREAM_END:
           case Z_NEED_DICT:
		break;

	   default:
		syslog(LOG_ERR, "compression failure\n");
		exit(5);
	}

	compbuf[0] = (Byte)complen;

	/*
	** write compressed netpertask struct, headed by one byte
	** with the size of the compressed struct
	*/
	if ( write(fd, compbuf, complen+1) < complen)
	{
		syslog(LOG_ERR, "write failure\n");
		exit(5);
	}
}

/*
** dummy handler for SIGHUP
*/
void
gethup(int sig)
{
}
